As a result of the human genome project sequenced we have the consensus sequence of all human chromsomes, as well as several other species. We say consensus sequence because every individual has a different sequence. But well over 99% is the same. 

Suppose you want to ask a questions such as: how many times does the sequence "ATG" appear on chromosome 11 ? Or what are the percentage of A,T,C and G on chromosome 7?

We can answer such question using Bioconductor tools. The human genome sequence is provided in the BSgenome.Hsapiens.UCSC.hg19 package. If you have not done so already please donwload and install this package. Note that it encodes 3 billion bases and is therefore a large package (over 800MB) so make time to download it especially if you have a slow internet connection.


```{r}
library(BSgenome.Hsapiens.UCSC.hg19)
BSgenome.Hsapiens.UCSC.hg19
```
We can access chromosome 11 like this:

```{r}
chr11seq <- BSgenome.Hsapiens.UCSC.hg19[["chr11"]]
```
Here, for example, is a segment of 25 bases starting  at base 1 million 


```{r}
subseq(chr11seq,start=10^6,width=25)

```

Frequencies of short sequences
2 points possible (graded)
Read the help file for the fuction countPattern and tell us which of the following sequences is most common on chromosome 11: "ATG", "TGA", "TAA", and "TAG"

Select one:



```{r}
countPattern("ATG",chr11seq)
```

```{r}

countPattern("TAA",chr11seq)
```

```{r}
countPattern("TAG",chr11seq)

```



```{r}
seqs<-c("ATG","TGA","TAA","TAG")
n<-sapply(seqs,function(sq){countPattern(sq,chr11seq)})
n
which.max(n)
```
# Nucleotide frequencies
1 point possible (graded)
Now we move to a question about chromosome 7. Read the help page for the function alphabetFrequency and use it to determine what percent of chromosome 7 is T,C,G, and A. Note that we have other letters. For example N, which represents positions that are not called, appears often.

What proportion are Cs (including counts of N in the total)

```{r}
chr7seq <- BSgenome.Hsapiens.UCSC.hg19[["chr7"]]
```



```{r}
alphabetFrequency(chr7seq,as.prob=TRUE)
```

```{r}
0.19901933+0.02378429
```
Locations of SNPs in humans
1 point possible (graded)
As explained in the video, many of the locations on the genome that are different across individual are single nucleotide polymorphisms (SNPs). This information is not on the human genome reference sequence. Instead, this information is stored in databases such as dbSNP. Bioconductor also gives you access to these database via the

package SNPlocs.Hsapiens.dbSNP144.GRCh37. Download and install this package. This is also a large package.

```{r}
if (!("SNPlocs.Hsapiens.dbSNP144.GRCh37" %in% rownames(installed.packages()))) {
   library(BiocInstaller)
   biocLite("SNPlocs.Hsapiens.dbSNP144.GRCh37")
   }
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
```




```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

if (!("SNPlocs.Hsapiens.dbSNP144.GRCh37" %in% rownames(installed.packages()))) {
BiocManager::install("SNPlocs.Hsapiens.dbSNP144.GRCh37")
   }

```
To see all the SNPs on, for example, chromosome 17 we can use the following commands


```{r}
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
snps144 = SNPlocs.Hsapiens.dbSNP144.GRCh37
s17 = snpsBySeqname(snps144, "17")
head(s17)


```

```{r}
s17["RefSNP_id"]
```


```{r}
s17 = snpsBySeqname(snps144, "17","rs73971683")

```

```{r}
## Lookup some RefSNP ids:
my_rsids <- c("rs73971683")
## Not run: 
snpsById(s17,my_rsids )  # error, rs7553394 not found



```

```{r}

s17[which(s17$RefSNP_id=="rs73971683")]
##or with dplyr
library(dplyr)
s17 %>% filter(RefSNP_id=="rs73971683") %>% select(pos)
```

```{r,eval=False}
install.packages("gwascat")

```
```{r}
if (!("gwascat" %in% rownames(installed.packages()))) {
BiocManager::install("gwascat")
   }
```

```{r}
library(gwascat)
data(ebicat37)
ebicat37
```

```{r}
ebicat37
```


```{r}

sort(table(mcols(ebicat37)[,"DISEASE/TRAIT"]),decreasing=TRUE)
```

```{r}

```

