```{r geter,eval=FALSE}
library(devtools)
install_github("genomicsclass/ERBS")
```

Once installed you can attach two object, one for each cell line, and view it's content:
```{r}
library(ERBS)
data(HepG2)  # cell line of liver origin
data(GM12878)  # immortalized B cell
HepG2  # locations of ER binding peaks
```

```{r}

```


Note that these object contain two main parts, the regions of the genome which we can extract with  `granges`:
```{r}
granges(HepG2)
```
and other information that we can obtain with either the `mcols` or `values` functions:
```{r}
values(HepG2) ##or mcols(HepG2)
```

<a name="basic"></a>
# Basic operations 

These object are of class `GRanges` defined in the `GenomicsRanges` package:
```{r}
class(HepG2)
```
There are a large number of tools available to use for this class and the best way to learn about them is through the vignette `browseVignettes("GenomicRanges")`

Here we start with some very basic example. Suppose these regions are ordered and we want to work with just the first 10. We can subset these objects as you would expect:
```{r}
granges( HepG2[1:10] )
```
Now what if we want to subset the object by chromosomes? We can access the chromosomes with `seqnames`. Note that not all organisms have chromosomes so Bioconductor uses a more genreal name:

```{r}
chr <- seqnames(HepG2)
chr
```
```{r}
as.character(chr)
```

Top 10
```{r}
HepG2[1:10,]

```
```{r}


```



Note that chr is not a factor or character vector as we would expect. Instead run-length encoding is used. The reason for this is that it provides a much more efficient way to store large granges, for example those obtained from stroing short reads. Note also that we have chromosome names that are not the standard chromosome names such as `chrUn_gl000247`. The human genome actually has some sequences that have not been mapped into one of the chromosomes and are given names like these. 

The Rle class behaves like factors in several useful ways. For example we can tabulate:


#Group Chomosome with frequency
```{r}
table(chr)[1:24]
```

We can also subset the object to data based on these. Here is the subset of regions on chr20:

```{r}
granges( HepG2[ chr == "chr20" ] )
```


#order by seqname then chromosome location

```{r}
x=HepG2[order(HepG2),] 
```

#SOrted
```{r}
seqnames(x)
```

